### PR TITLE
android virtualkey with corrector/completion

### PR DESCRIPTION
**Pull request by Gorash**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web_editor/static/src/js/wysiwyg/plugin/keyboard.js [hunk 1]
```diff
@@ -998,7 +994,47 @@ var KeyboardPlugin = AbstractPlugin.extend({
...
+            if (parentContenteditable) {
+                $parent.attr('contenteditable', parentContenteditable);
+            } else {
+                $parent.removeAttr('contenteditable');
+            }
+            $(accentPlaceholder).remove();
+            callback(text, baseRange);
+            this._onAccentedAndVirtualKeyCallback = null;
+        };
+        this._onAccentedAndVirtualKeyTimer = setTimeout(this._onAccentedAndVirtualKeyCallback.bind(this));
```
Comment (VincentSchippefilt) [inline addons/web_editor/static/src/js/wysiwyg/plugin/keyboard.js]: This setTimeout seems dangerous, the event stack might be flooded by touch/scroll events

File: addons/web_editor/static/src/js/wysiwyg/plugin/keyboard.js [hunk 2]
```diff
@@ -1230,38 +1292,86 @@ var KeyboardPlugin = AbstractPlugin.extend({
...
+     */
+    _onAndroidVirtualKeyboard: function (e) {
+        var self = this;
+        e.preventDefault();
+
+        if (!e.originalEvent.isComposing) {
+            return;
+        }
+
+        this._onAccentedAndVirtualKey(function (text, range) {
```
Comment (VincentSchippefilt) [inline addons/web_editor/static/src/js/wysiwyg/plugin/keyboard.js]: this callback is big enough that it deserves a name and documentation ;)

File: addons/web_editor/static/src/js/wysiwyg/plugin/keyboard.js [hunk 3]
```diff
@@ -1221,46 +1190,125 @@ var KeyboardPlugin = AbstractPlugin.extend({
...
+            range.sc = range.ec = point.node;
+            range.so = range.eo = point.offset;
+            range.select();
+        }
+
+        if (text.length) {
+            this.context.invoke('HelperPlugin.insertTextInline', text);
+            var range = this.context.invoke('editor.createRange');
+            var textContent = range.sc.textContent;
+            function fuckGoogleKeyboardWhoInsertCharWithoutEvent () {
```
Comment (VincentSchippefilt) [inline addons/web_editor/static/src/js/wysiwyg/plugin/keyboard.js]: fuckGoogleKeyboardWho**m**Insert**s**CharWithoutEvent

Comment (pparidans) [reply addons/web_editor/static/src/js/wysiwyg/plugin/keyboard.js]: :+1:

Comment (Zinston) [reply addons/web_editor/static/src/js/wysiwyg/plugin/keyboard.js]: fuckGoogleKeyboard**Which**InsertsCharWithoutEvent

Comment (Zinston) [reply addons/web_editor/static/src/js/wysiwyg/plugin/keyboard.js]: or fuckGoogleKeyboard**That**InsertsCharWithoutEvent

Comment (Zinston) [reply addons/web_editor/static/src/js/wysiwyg/plugin/keyboard.js]: Now, actuallyt, saying it out loud I realize "that" would be better than "which".

**Comment by Gorash** [comment]
@pparidans ping

**Comment by pparidans** [comment]
Ping @Gorash @Zinston for the review.

**Comment by pparidans** [comment]
After some re-testing, I encountered the following issues :

- Firefox on Windows/MacOS/Ubuntu :
  * you can't introduce new characters after pressing enter key.
- Safari on iOS :
  * autocorrect disabled when you modify an existing word.
  * <s>when you autocorrect the first word of the line a space is added in front of it.</s> _Weirdly I can't reproduce this issue._
- Safari on MacOS :
  * autocorrect disabled when you modify an existing word.
- Internet Explorer 11 on Windows :
  * you can't introduce new characters after pressing enter key.

**Comment by Gorash** [comment]
For the autocorrect I say in advance, I will disable it if possible. This is not essential and the Safari system does not follow the standards and the corrector does not update for some unknown reason. I will look at this separately. I will see for the enter.

**Comment by pparidans** [comment]
There is also a massive "glitch" on iOS: the view scroll up and down ("bounce") when pressing enter or backspace keys.

PS: I added this point to the previous comment for clarity

**Comment by Gorash** [comment]
@pparidans ping ;)

**Comment by adr-odoo** [comment]
@robodoo r+

**Comment by adr-odoo** [comment]
@robodoo r-

**Comment by adr-odoo** [comment]
@robodoo rebase-ff

**Comment by adr-odoo** [comment]
@robodoo r+

**Pull request closed without merging**